keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,711 @@
+discussion and make progress. -->
+
+- Accommodating new OpenAPI v3 fields outside the Schema object that do not
+  affect validity of the spec. These are nice to have fields that can be added
+  later.
+- Consumers of the OpenAPI (eg: Server Side Apply, client-go, kubectl, etc) will
+  eventually need to be updated to consume v3. This is outside the scope of this
+  KEP.
+
+
+## Proposal
+
+<!-- This is where we get down to the specifics of what the proposal actually
+is. This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?. The
+"Design Details" section below is for the real nitty-gritty. -->
+
+The proposal is to publish OpenAPI v3 schemas at the
+`/openapi/v3/apis/{group}/{version}` endpoint for all resources. Instead of
+aggregating everything into one schema, we will publish separate schemas per
+resource group-version. The aggregated schema will not be published at the
+openapi/v3 endpoint and the endpoint will instead be used for discovery
+purposes. For users who still need to retrieve the entire OpenAPI spec (not
+recommended), we will offer a client side utility to aggregate all the
+endpoints.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Future Work
+
+After OpenAPI v3 is implemented, we can start working on updating clients to
+consume OpenAPI v3. A separate KEP will be published for that.
+
+### Risks and Mitigations
+
+Aggregation for OpenAPI v2 already consumes a non-negligible amount of
+resources, naively adding in an aggregation for v3 will double the workload when
+the OpenAPI needs to be re-aggregated. Splitting by group-version partially
+mitigates this problem because when a resource is changed, only a small subset
+of groups will need to be reaggregated compared to the current behavior of
+re-aggregating everything.
+
+## Design Details
+
+<!-- This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them. -->
+
+### Paths
+
+The overarching `/openapi/v3` endpoint will contain the list of paths (groups)
+available and serve as a discovery endpoint. Clients can then choose the
+group(s) to fetch and send additional requests.
+
+/openapi/v3
+
+```json
+{
+   "Paths" : [
+      "api",
+      "api/v1",
+      "apis",
+      "apis/admissionregistration.k8s.io",
+      "apis/apiextensions.k8s.io",
+      "apis/apps",
+      "apis/authentication.k8s.io",
+      ...
+   ]
+}
+```
+
+Based on the provided group, clients can then request `openapi/v3/apis/apps/v1`,
+`/openapi/v3/apis/networking.k8s.io/v1` and etc. These leaf node specs are self
+contained OpenAPI v3 specs and include all referenced types.
```

Comment:
I know it is late, but why don't we publish under `/apis/networking.k8s.io/v1/openapi`? It will remove all wiring into the aggregator.

Reply:
cc @liggitt @deads2k for opinions.

This feels like a natural step and considerably simplifies the wiring and architecture within kube-apiserver. Basically apiextensions-apiserver will serve the spec directly. kube-aggregator will stop internally downloading the json blob and caching it in the aggregation layer. In the same way, kube-aggregator will not need special handling for openapi because /apis/foo/v1 is proxied to the aggregated apiserver already. That was the original reason why we switch to per-group-version specs here in the enhancement.

Reply:
`/apis/networking.k8s.io/v1/openapi` conflicts with the endpoints for an `openapi` resource

Reply:
(and matches the pattern for a resource URL instead of a non-resource URL... it would get authorized as `verb=list,resource=openapi,group=networking.k8s.io,scope=cluster-wide`

Reply:
We can still proxy the requests to the apiextensions server even if the URL is different?

Reply:
We can proxy them, yes. Am fine with that approach. In the implementation PR, please make sure we avoid coupling apiextensions-apiserver with kube-aggregator (the downloader cames to mind). Rather serve the per-gv specs directly from the crd handler.